### DISCUSSION TITLE
Handling MacOS/Windows App Updates (Captive Runtime)

### DISCUSSION THREAD

**Discussion by KarlSchumann** [airsdk/Adobe-Runtime-Support#1294] [Q&A]
We have a desktop application that we publish for Mac and Windows and previously relied on the shared runtime where we could leverage the AIR Update framework for managing application updates, but now we are switching to the captive runtime and it doesn't seem as straight forward so curious how others are handling.

In the short term I think we will just be directing users to redownload the app and manually install it to handle an application level update as they aren't as common for us. But ideally longer term we would have something within the app itself or at launch that would handle application updates in a more automatic way similar to how the experience we used to have using the shared runtime.

Any info others could share with their experience would be greatly appreciated.
Thanks.

**Comment by MalacTheLittle** [comment]
Here are some toughts how you can solve this...

The beauty of non-AOT compiled apps is that you can do a lot of unusual stuff with it. Here's an example how I made application that we distributed over the USB stick and updates where made just by downloading new .swf file and restart app by itself.
I made two bundle apps (bundle = with included AIR runtime): first one is starter (I'll call it here outter app), and second one is actual application (inner app). Inner app is, of course, located inside folder of outter app, so when user starts outter app, outter app just runs .exe of inner app and closes itself. The reason for that approach is that if you want to update AIR runtime of inner app, you can not do it for app itself. So, you need other (outter) app to do it for inner app. Also, if you want to update AIR for outter app, you can do it from inner app.

You can always update .swf of your app (if it's not installed in restricted systems folders) without using other app because .exe file just reads .swf and closes stream to it. But sometimes you need to update AIR runtime and .exe file (I think .dll file is occupied by system also), and than you need other app for that action.

At the end, there were two reasons for this approach:
1. to be able to update AIR runtime of inner app
2. to be able to restart app after update of .swf.

Now, the problem with this approach is that if user installs this kind of app inside Program Files folder, system doesn't allow you to write in apps folder. Not sure if it allows to write in apps subfolders, but you can on the first run of the outter app copy all files of inner app to outters app Storage folder so outter app runs inner app from appStorage folder. That way you can update .swf and AIR runtime with no restrictions :)

I'm not sure if this approach is usable on MacOS, but I don't see a reason not to be.

Since Harman is encouraging developers to use bundle apps for distribution, it's true that we are faced with bigger update installs, but I find it as better end-users experience since they don't need to update app and AIR runtime so feeling is more as AOT for end-users. And installers made with WixToolset on Windows and with Packages on MacOS makes it as real AOT apps :)

Hope this helps you... :)

**Reply by KarlSchumann** [reply]
Thank you very much for the detailed response, that is very helpful. I definitely that like with the captive runtime that we no longer need to have the end-user first download/install the runtime, but tough to beat the convenience that installation and updates were built-in with that approach. I believe I have packaging/signing/notarizing figured out for the Mac side of things and next will be making a proper installer for Windows, have you had good success with the WixToolset? That was on my list to look into.

Appreciate you taking the time to share your experience.

**Reply by manishpatel0404** [reply]
Another easy to implement option for a Windows installer is [Inno Setup](https://jrsoftware.org/isinfo.php). It is well documented and customizable through scripting.

You can perform application level updates using AS3 NativeProcess with your installer (.exe) file. For this, your app would essentially download the installer to a temp folder, use NativeProcess to invoke the installer before exiting itself and let the installer take over. This way, you don't have to have users download the installer and run it themselves.

For updates (Main swf, other assets) that do not require the entire application to be updated, you can create a stub projector that contains logic to check which file(s) need to be updated, including the Main swf, download the required updates to the user's application storage folder and load the app's Main swf. Package only the stub projector files along with the AIR captive runtime files + ANEs, etc using an installer. This way, you can update your app at will and also reduce the size of the installer.

Hope this helps.

**Reply by MalacTheLittle** [reply]
@manishpatel0404 I wish I knew about Inno Setup few months ago when I was faced with native packaging for the first time because it was quite an adventure to make it work :)
But, when I look at it from todays perspective, I don't actually regret it since WixToolset is quite powerfull for making .msi installers which are executed by Windows (Microsoft) Installer. Inno Setup, as I could find in short time, doesn't support packaging to .msi packages, @manishpatel0404 please correct me if I'm wrong. I emphasize .msi installer because you can run it from your application just by calling YourInstallerMSI_File`.openWithDefaultApplication()` :) Same thing happens on MacOS when you package your installer in .pkg package. This thing is more than usefull on both OS's. 
On Windows, if you don't exit application after calling `openWithDefaultApplication()`, installer will pause and ask user to close application before proceeding :)

@KarlSchumann Yes, I had good success with WixToolset and with Packages app on MacOS. I can point you to small tool that will ease your pain with WixToolset journey because it's quite bad documented. Tool is called [WixEdit](https://wixedit.github.io/). Altough, there's a great book with all the info you need for building an installer ([WiX Cookbook](https://www.perlego.com/book/117874/wix-cookbook-pdf)). I didn't had these needs like you have, but I believe you can make update installer with just files you need to update...
I can also share my .wxs setup file with you if you'll need.

**Reply by KarlSchumann** [reply]
@manishpatel0404 Thank you for the suggestion, I will take a look at that as well. Right now for Mac we are simply creating a .dmg using [node-appdmg](https://github.com/LinusU/node-appdmg) and still looking into options for Windows side. We have created our own updates for some assets within the app that are more frequently changing which has worked well, but now investigating options for updating the application itself as well as AIR as we transition to the captive runtime.

**Comment by 42retfa** [comment]
Hello,

Here is an old Adobe tutorial (no longer hosted by adobe) on how to generate a windows installer using WixEdit: [Generating a windows installer for an AIR captive runtime application](https://github.com/airsdk/Adobe-Runtime-Support/files/7416953/GENERATING.A.WINDOWS.INSTALLER.FOR.YOUR.AIR.CAPTIVE.RUNTIME.APPLICATION.pdf)

Vincent.
